File: LeetCode/1887_ReductionOperationsToMakeTheArrayElementsEqual/lc1887.cpp
#include <algorithm>
#include <cassert>
#include <vector>

class Solution {
public:
  int reductionOperations(std::vector<int>& nums) {
    std::sort(nums.begin(), nums.end());
    int ans = 0;
    int count_of_different_numbers = 0;
    for (int i = 1; i < nums.size(); i++) {
      if (nums[i] > nums[i-1]) {
        ++count_of_different_numbers;
      }
      ans += count_of_different_numbers;
    }

    return ans;
  }
};

int main() {
  Solution s;
  {
    std::vector<int> nums = {5, 1, 3};
    assert(s.reductionOperations(nums) == 3);
  }
  {
    std::vector<int> nums = {1, 1, 1};
    assert(s.reductionOperations(nums) == 0);
  }
  {
    std::vector<int> nums = {1, 1, 2, 2, 3};
    assert(s.reductionOperations(nums) == 4);
  }
}

File: LeetCode/1887/python/lc1887.py
from typing import List

# Runtime complexity: O(sort)
# Auxiliary space complexity: O(sort)
# Subjective level: medium+ (coming up with the actual solution to this was a bit tricky,
#                   as duplicate numbers are allowed in the input data).
# Solved on: 2026-02-27
class Solution:
    def reductionOperations(self, nums: List[int]) -> int:
        L = len(nums)
        nums.sort()
        ans = 0
        for i in range(1, L):
            if nums[i] != nums[i-1]:
                ans += L - i
        return ans
